libs/wingc/src/inflight_transform.rs

```diff
@@ -0,0 +1,277 @@
+			ExprKind::FunctionClosure(func_def) => {
+				self.inflight_counter += 1;
+
+				let resource_name = Symbol {
+					name: format!("$Inflight{}", self.inflight_counter),
+					span: expr.span.clone(),
+				};
+				let handle_name = Symbol {
+					name: "handle".to_string(),
+					span: expr.span.clone(),
+				};
+				let parent_this_name = Symbol {
+					name: PARENT_THIS_NAME.to_string(),
+					span: expr.span.clone(),
+				};
+				let this_name = Symbol {
+					name: "this".to_string(),
+					span: expr.span.clone(),
+				};
+
+				let resource_type_annotation = TypeAnnotation {
+					kind: TypeAnnotationKind::UserDefined(UserDefinedType {
+						root: resource_name.clone(),
+						fields: vec![],
+					}),
+					span: expr.span.clone(),
+				};
+
+				let resource_fields: Vec<ClassField> = vec![];
+				let resource_init_params: Vec<FunctionParameter> = vec![];
+
+				let new_func_def = if self.is_inside_resource {
+					// transform all occurrences of "this" into "__parent_this"
+					let mut this_transform = ThisTransformer;
+					this_transform.fold_function_definition(func_def)
+				} else {
+					func_def
+				};
+
+				// resource_def :=
+				// ```
+				// resource {
+				//   init(<resource_init_params>) {<resource_init_body>}
+				//   inflight handle() {
+				//     <transformed_def>
+				//   }
+				// }
+				// ```
+				let resource_def = Stmt {
+					kind: StmtKind::Class(Class {
+						name: resource_name.clone(),
+						is_resource: true,
+						initializer: Initializer {
+							signature: FunctionSignature {
+								parameters: resource_init_params,
+								return_type: Some(Box::new(resource_type_annotation.clone())),
+								phase: Phase::Preflight,
+							},
+							statements: Scope::new(vec![], expr.span.clone()),
+							span: expr.span.clone(),
+						},
+						fields: resource_fields,
+						implements: vec![],
+						parent: None,
+						methods: vec![(handle_name.clone(), new_func_def)],
+						inflight_initializer: None,
+					}),
+					idx: 1,
+					span: expr.span.clone(),
+				};
+
+				// return_resource_instance :=
+				// ```
+				// return new <resource_name>();
+				// ```
+				let return_resource_instance = Stmt {
+					kind: StmtKind::Return(Some(Expr::new(
+						ExprKind::New {
+							class: resource_type_annotation,
+							arg_list: ArgList {
+								named_args: IndexMap::new(),
+								pos_args: vec![],
+							},
+							obj_id: None,
+							obj_scope: None,
+						},
+						expr.span.clone(),
+					))),
+					idx: 2,
+					span: expr.span.clone(),
+				};
+
+				// make_resource_body :=
+				// ```
+				// {
+				//   <parent_this_def>? // only if we are inside a resource
+				//   <resource_def>
+				//   <return_resource_instance>
+				// }
+				// ```
+				let make_resource_body = if self.is_inside_resource {
+					// parent_this_def :=
+					// ```
+					// let __parent_this = this;
+					// ```
+					let parent_this_def = Stmt {
+						kind: StmtKind::VariableDef {
```

Comment:
```suggestion
						kind: StmtKind::Let {
```

We should rename this at some point